File: 943-SumOfSubarrayMinimums/943-SumOfSubarrayMinimums.cpp
// Last updated: 7/12/2025, 11:54:17 PM
class Solution {
public:
    int sumSubarrayMins(std::vector<int>& arr) {
        const int MOD = 1000000007;
        int res = 0;
        std::stack<std::pair<int, int>> stack;  // index, num

        for (int i = 0; i < arr.size(); ++i) {
            while (!stack.empty() && arr[i] < stack.top().second) {
                auto [j, m] = stack.top();
                stack.pop();
                int left = j - (stack.empty() ? -1 : stack.top().first);
                int right = i - j;
                res = (res + static_cast<long long>(m) * left * right) % MOD;
            }
            stack.push({i, arr[i]});
        }

        while (!stack.empty()) {
            auto [j, n] = stack.top();
            stack.pop();
            int left = j - (stack.empty() ? -1 : stack.top().first);
            int right = arr.size() - j;
            res = (res + static_cast<long long>(n) * left * right) % MOD;
        }

        return res;
    }
};

File: 943-SumOfSubarrayMinimums/943-SumOfSubarrayMinimums.py
# Last updated: 7/12/2025, 11:54:21 PM
class Solution:
    def sumSubarrayMins(self, arr: List[int]) -> int:
        MOD = 10 ** 9 + 7
        res = 0
        stack = [] # index, num


        for i,n in enumerate(arr):
            while stack and n < stack[-1][1]:     
                j, m = stack.pop()
                left = j - stack[-1][0] if stack else j + 1
                right = i - j
                res = (res + m * left * right) % MOD
            stack.append((i,n))

        for i in range(len(stack)):
            j, n = stack[i]
            left = j - stack[i - 1][0] if i > 0 else j + 1
            right = len(arr) - j 
            res = (res + n * left * right) % MOD

        return res
